edb/errors/compiler.py

```diff
@@ -0,0 +1,56 @@
+from edb import errors
+from edb import schema
+
+
+class BadTypeObject(errors.InvalidReferenceError):
+    def __init__(self, obj_id, obj_type, target_type,
+                 obj_name=None, context=None):
+        self._obj_id = obj_id
+        self._obj_type = obj_type.get_schema_class_displayname()
+        self._obj_name = obj_name
+        self._target_type = target_type.get_schema_class_displayname()
+        self._context = context
+        super().__init__(
+            f'schema object {obj_id!r} exists, but is not '
+            f'{target_type.get_schema_class_displayname()}')
+
+    def update(self, obj_name=None, source_context=None, statement=None):
```

Comment:
I'd prefer not to create exceptions outside of `edb.errors`. We started with this and it took us quite some time to clean everything up. I'd suggest generalizing this and implementing directly in the `edb.errors.base.EdgeDBError` class.

@elprans thoughts?

Reply:
Technically, it's not a new error type since it's reduced back to `InvalidReferenceError`, but I'm a bit worried about the error "update" pattern.  Perhaps passing the necessary context to `schema.get` is cleaner?

Reply:
> Technically, it's not a new error type since it's reduced back to InvalidReferenceError

But that's my point. I don't want us to have 10s of intermediate error types reducing to `edb.errors` eventually. This complexity is unnecessary and is a sure road to spaghetti code.

Reply:
So here is the original traceback, marked with places where we add context:
```
      File "/work/edb/schema/types.py", line 1987, in _create_begin
        schema = super()._create_begin(schema, context)
      File "/work/edb/schema/inheriting.py", line 595, in _create_begin
        schema = super()._create_begin(schema, context) <<< HERE WE ADD STATEMENT
      File "/work/edb/schema/inheriting.py", line 44, in _create_begin
        schema = super()._create_begin(schema, context)
      File "/work/edb/schema/delta.py", line 2119, in _create_begin
        self.validate_create(schema, context)
      File "/work/edb/schema/types.py", line 2017, in validate_create
        bases = self.get_resolved_attribute_value(
      File "/work/edb/schema/delta.py", line 1804, in get_resolved_attribute_value
        value = self._resolve_attr_value(
      File "/work/edb/schema/delta.py", line 361, in _resolve_attr_value
        val = v.resolve(schema) <<< HERE WE ADD CONTEXT (code span)
      File "/work/edb/schema/types.py", line 473, in resolve
        return schema.get(
      File "/work/edb/schema/schema.py", line 339, in get
        return self.get_generic(
      File "/work/edb/schema/schema.py", line 1690, in get_generic
        return self._base_schema.get(
      File "/work/edb/schema/schema.py", line 339, in get
        return self.get_generic(
      File "/work/edb/schema/schema.py", line 1238, in get_generic
        obj = self._get(name, <<< HERE WE ADD REAL NAME OF THE OBJECT
      File "/work/edb/schema/schema.py", line 962, in _get
        result = getter(self, fqname)
      File "/work/edb/schema/schema.py", line 1232, in getter
        obj = schema.get_by_id(obj_id, type=type, default=None)
      File "/work/edb/schema/schema.py", line 1174, in get_by_id
        raise errors.InvalidReferenceError(
    edb.errors.InvalidReferenceError: schema object UUID('00000000-0000-0000-0000-000000000101') exists, but is not object type
```

Roughly:
1. Statement layer should be brought 13 layers below
2. Context (code span) should be brought 8 layers below
3. Real name should be brought 3 layers below

This roughtly translates as:
1. 13 functions should have a new argument
2. 8 function should have 2 new arguments
3. 3 functions should have 3 new arguments

Looking at it now I think it's probably possible to merge 2-3 into one point, but this doesn't improve situation to a manageable level.

While I understand that you need to wrap your head around that, but generally providing more context to the error is not only hugely used in Rust, it's also quite often used in python, like when you do:
```
except KeyError as e:
    raise FormValidationError(...)
```
... you essentially add more information to the error. This is also cheaper on success  part (comparing to passing all that context via args through call chain), and probably even on error case too.

We can bikeshed about how specifically we add that (i.e. method is called `update` rather than `context` like in Rust, because `context` term is used for source code position, etc.).

> I'd suggest generalizing this and implementing directly in the edb.errors.base.EdgeDBError class.

I'm not sure that `update()` method at the base error is very good, as many errors could have different context information needed. Please share your thoughts on how this should be generalized.

> But that's my point. I don't want us to have 10s of intermediate error types reducing to edb.errors eventually. This complexity is unnecessary and is a sure road to spaghetti code.

1. We have fields in this error object that can't be serialized efficiently and are useless outside of the compiler anyway
2. Without fields, I don't think it's viable to parse that information from the text message

Generally I don't see any problem in compiler-only exceptions. We already have exceptions, like `KeyError`, `ValueError`, etc. that should not leave compiler boundary and should be translated into a valid client error.

Reply:
The passing of a bunch of contextual information trough the stack isn't new in this piece of code, which is why we pass `DeltaContext` everywhere. Here we just need to see how we can associate the extra info with that.  For example, the source span is already present in the `source_context` attribute of the current DDL command, which you can get from the context.  Likewise we can retain the reference to the original AST there also.

Thus, no new arguments would be needed, except at the `schema.py` level, where the call stack is fairly shallow and there is already precedent in other similar arguments.

Reply:
> The passing of a bunch of contextual information trough the stack isn't new in this piece of code, which is why we pass DeltaContext everywhere.

It stops propagating at one layer before resolve, where I add code span. Which is fixable, but...

> For example, the source span is already present in the source_context

This is not that source context. We need to pick specific name not the whole statement.

> Thus, no new arguments would be needed, except at the schema.py level, where the call stack is fairly shallow

I still see 8 layers above resolve. And CommandContext is not enough, we should make a more specific context object for that, that includes specific name.

> and there is already precedent in other similar arguments.

This is specifically frightening. Just a single simple hint needs another argument in the core and otherwise simple function `get_by_id`. More hints -- more places like that.

----

That said, it looks like we can aggregate all three `update`'s into a single error transformation at around `resolve` call, and we can use `CommandContext` instead of statement itself when figuring out hint. But I'm not convinced that we should pass the whole context forward.